src/Init/Data/List/Basic.lean

```diff
@@ -276,54 +385,93 @@ theorem mem_append_of_mem_right {b : α} {bs : List α} (as : List α) : b ∈ b
   | nil  => simp [h]
   | cons => apply Mem.tail; assumption
 
-def eraseDupsAux {α} [BEq α] : List α → List α → List α
+/-- `O(|l|^2)`. Erase duplicated elements in the list.
+Keeps the first occurrence of duplicated elements.
+* `eraseDups [1, 3, 2, 2, 3, 5] = [1, 3, 2, 5]`
+-/
+def eraseDups {α} [BEq α] (as : List α) : List α :=
+  loop as []
+where
+  loop : List α → List α → List α
   | [],    bs => bs.reverse
   | a::as, bs => match bs.elem a with
-    | true  => eraseDupsAux as bs
-    | false => eraseDupsAux as (a::bs)
+    | true  => loop as bs
+    | false => loop as (a::bs)
 
-def eraseDups {α} [BEq α] (as : List α) : List α :=
-  eraseDupsAux as []
-
-def eraseRepsAux {α} [BEq α] : α → List α → List α → List α
+/--
+`O(|l|)`. Erase repeated adjacent elements. Keeps the first occurrence of each run.
+* `eraseReps [1, 3, 2, 2, 2, 3, 5] = [1, 3, 2, 3, 5]`
+-/
+def eraseReps {α} [BEq α] : List α → List α
+  | []    => []
+  | a::as => loop a as []
+where
+  loop {α} [BEq α] : α → List α → List α → List α
   | a, [], rs => (a::rs).reverse
   | a, a'::as, rs => match a == a' with
-    | true  => eraseRepsAux a as rs
-    | false => eraseRepsAux a' as (a::rs)
+    | true  => loop a as rs
+    | false => loop a' as (a::rs)
 
-/-- Erase repeated adjacent elements. -/
-def eraseReps {α} [BEq α] : List α → List α
-  | []    => []
-  | a::as => eraseRepsAux a as []
+/--
+`O(|l|)`. `span p l` splits the list `l` into two parts, where the first part
+contains the longest initial segment for which `p` returns true
+and the second part is everything else.
 
-@[specialize] def spanAux (p : α → Bool) : List α → List α → List α × List α
+* `span (· > 5) [6, 8, 9, 5, 2, 9] = ([6, 8, 9], [5, 2, 9])`
+* `span (· > 10) [6, 8, 9, 5, 2, 9] = ([6, 8, 9, 5, 2, 9], [])`
+-/
+@[inline] def span (p : α → Bool) (as : List α) : List α × List α :=
+  loop as []
+where
+  @[specialize] loop : List α → List α → List α × List α
   | [],    rs => (rs.reverse, [])
   | a::as, rs => match p a with
-    | true  => spanAux p as (a::rs)
+    | true  => loop as (a::rs)
     | false => (rs.reverse, a::as)
 
-@[inline] def span (p : α → Bool) (as : List α) : List α × List α :=
-  spanAux p as []
+/--
+`O(|l|)`. `groupBy R l` splits `l` into chains of elements
+such that adjacent elements are related by `R`.
 
-@[specialize] def groupByAux (eq : α → α → Bool) : List α → List (List α) → List (List α)
-  | a::as, (ag::g)::gs => match eq a ag with
-    | true  => groupByAux eq as ((a::ag::g)::gs)
-    | false => groupByAux eq as ([a]::(ag::g).reverse::gs)
+* `groupBy (·==·) [1, 1, 2, 2, 2, 3, 2] = [[1, 1], [2, 2, 2], [3], [2]]`
+* `groupBy (·<·) [1, 2, 5, 4, 5, 4, 1] = [[1, 2, 5], [4, 5], [4], [1]]`
+-/
+@[specialize] def groupBy (R : α → α → Bool) : List α → List (List α)
+  | []    => []
+  | a::as => loop as [[a]]
+where
+  @[specialize] loop : List α → List (List α) → List (List α)
+  | a::as, (ag::g)::gs => match R ag a with
```

Comment:
Note that I reversed the arguments `ag` and `a` here. The original code named the variable `eq` so it was presumably assuming symmetry, but in the general case it finds chains which are connected by the relation `R`, so this argument order makes more sense (otherwise it is `(·>·)` that finds increasing runs in the doc example).